test/gateway/eip20_cogateway/progress_revert_redeem.js

```diff
@@ -0,0 +1,385 @@
+      "Receipt status is unsuccessful",
+    );
+
+  });
+
+  it('should emit `RedeemReverted` event ', async function () {
+
+    await cogateway.setStorageRoot(revertRedeemParams.blockHeight, revertRedeemParams.storageRoot);
+
+    let tx = await cogateway.progressRevertRedeem(
+      revertRedeemParams.messageHash,
+      revertRedeemParams.blockHeight,
+      revertRedeemParams.rlpParentNodes,
+    );
+
+    let event = EventDecoder.getEvents(tx, cogateway);
+    let eventData = event.RedeemReverted;
+
+    assert.isDefined(
+      event.RedeemReverted,
+      'Event `RedeemReverted` must be emitted.',
+    );
+
+    assert.strictEqual(
+      eventData._messageHash,
+      revertRedeemParams.messageHash,
+      `Message hash from the event must be equal to ${revertRedeemParams.messageHash}.`,
+    );
+
+    assert.strictEqual(
+      eventData._redeemer,
+      revertRedeemParams.redeemer,
+      `Redeemer address from the event must be equal to ${revertRedeemParams.redeemer}.`,
+    );
+
+    assert.strictEqual(
+      eventData._redeemerNonce.eq(revertRedeemParams.nonce),
+      true,
+      `Redeemer nonce from the event must be equal to ${revertRedeemParams.nonce}.`,
+    );
+
+    assert.strictEqual(
+      eventData._amount.eq(revertRedeemParams.amount),
+      true,
+      `Redeem amount from the event must be equal to ${revertRedeemParams.amount}.`,
+    );
+
+  });
+
+  it('should burn bounty and penalty amount ', async function () {
+
+    let burnerAddress = await cogateway.burner.call();
+
+    let cogatewayInitialTokenBalance = await Utils.getBalance(cogateway.address);
+    let redeemerInitialTokenBalance = await Utils.getBalance(revertRedeemParams.redeemer);
+    let burnerInitialTokenBalance = await Utils.getBalance(burnerAddress);
+
+    await cogateway.setStorageRoot(revertRedeemParams.blockHeight, revertRedeemParams.storageRoot);
+
+    await cogateway.progressRevertRedeem(
+      revertRedeemParams.messageHash,
+      revertRedeemParams.blockHeight,
+      revertRedeemParams.rlpParentNodes,
+    );
+
+    let cogatewayFinalTokenBalance = await Utils.getBalance(cogateway.address);
+    let redeemerFinalTokenBalance = await Utils.getBalance(revertRedeemParams.redeemer);
+    let burnerFinalTokenBalance = await Utils.getBalance(burnerAddress);
+
+    assert.strictEqual(
+      cogatewayFinalTokenBalance.eq(cogatewayInitialTokenBalance.sub(bountyAmount.muln(2.5))),
+      true,
+      `CoGateway balance must decrease by ${bountyAmount.muln(2.5).toString(10)}.`,
+    );
+
+    assert.strictEqual(
+      redeemerInitialTokenBalance.eq(redeemerFinalTokenBalance),
+      true,
+      `Redeemer balance must not change.`,
+    );
+
+    assert.strictEqual(
+      burnerFinalTokenBalance.eq(burnerInitialTokenBalance.add(bountyAmount.muln(2.5))),
+      true,
+      `Burner balance must increase by ${bountyAmount.muln(2.5).toString(10)}.`,
+    );
+
+  });
+
+  it('should transfer redeem amount to redeemer ', async function () {
```

Comment:
Why does the redeemer balance change in this test but not in the previous one? It is not obvious to me 😞

Reply:
In previous tests base token(Eth balance) doesn't change. The base token is only used for penalty and bounty. 
But here redeemer ERC20 balance should be returned as redeem revert is progressed.

Reply:
But it is in both cases `utilityToken.balanceOf(revertRedeemParams.redeemer);` 🧐
Please explain in the code as well.

Reply:
I didn't understand your point. Let's discuss this in a call.